server/api/notification.go

```diff
@@ -0,0 +1,301 @@
+	http.Handler
+	Configure(apiConfig Config)
+	Quit()
+}
+
+type notificationHandler struct {
+	Config
+	incoming   chan *remote.Notification
+	queue      chan *remote.Notification
+	queueSize  int
+	configChan chan Config
+	quit       chan bool
+}
+
+func NewNotificationHandler(apiConfig Config) NotificationHandler {
+	h := &notificationHandler{
+		Config:     apiConfig,
+		incoming:   make(chan (*remote.Notification)),
+		queue:      make(chan (*remote.Notification), maxQueueSize),
+		configChan: make(chan (Config)),
+		quit:       make(chan (bool)),
+	}
+	go h.work()
+	return h
+}
+
+func (h *notificationHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	notifications := h.Remote.HandleNotification(w, req)
+	for _, n := range notifications {
+		h.incoming <- n
+	}
+}
+
+func (h *notificationHandler) Configure(apiConfig Config) {
+	h.configChan <- apiConfig
+}
+
+func (h *notificationHandler) Quit() {
+	h.quit <- true
+}
+
+func (h *notificationHandler) work() {
+	for {
+		select {
+		case n := <-h.incoming:
+			if h.queueSize >= maxQueueSize {
+				h.Logger.LogError(
+					fmt.Sprintf("Notification queue full (%v), dropped notification", h.queueSize))
+				continue
+			}
+			h.queueSize++
+			h.queue <- n
+
+		case n := <-h.queue:
+			h.queueSize--
+			err := h.processNotification(n)
+			if err != nil {
+				h.Logger.LogInfo("Failed to process notification: "+err.Error(),
+					"SubsriptionID", n.SubscriptionID)
+			}
+
+		case apiConfig := <-h.configChan:
+			h.Config = apiConfig
+
+		case <-h.quit:
+			return
+		}
+	}
+}
+
+func (h *notificationHandler) processNotification(n *remote.Notification) error {
+	sub, err := h.SubscriptionStore.LoadSubscription(n.SubscriptionID)
+	if err != nil {
+		return err
+	}
+	creator, err := h.UserStore.LoadUser(sub.MattermostCreatorID)
+	if err != nil {
+		return err
+	}
+	if sub.Remote.ID != creator.Settings.EventSubscriptionID {
+		return errors.New("Subscription is orphaned")
+	}
+	if sub.Remote.ClientState != "" && sub.Remote.ClientState != n.ClientState {
+		return errors.New("Unauthorized webhook")
+	}
+
+	n.Subscription = sub.Remote
+	n.SubscriptionCreator = creator.Remote
+
+	var client remote.Client
+	if !n.RecommendRenew || n.IsBare {
+		client = h.Remote.NewClient(context.Background(), creator.OAuth2Token)
+	}
+
+	if n.RecommendRenew {
+		var renewed *remote.Subscription
+		renewed, err = client.RenewSubscription(n.SubscriptionID)
+		if err != nil {
+			return err
+		}
+
+		storedSub := &store.Subscription{
+			Remote:              renewed,
+			MattermostCreatorID: creator.MattermostUserID,
+			PluginVersion:       h.Config.PluginVersion,
+		}
+		err = h.SubscriptionStore.StoreUserSubscription(creator, storedSub)
+		if err != nil {
+			return err
+		}
+		h.Logger.LogDebug("Renewed user subscription",
+			"MattermostUserID", creator.MattermostUserID,
+			"SubsriptionID", n.SubscriptionID)
+	}
+
+	if n.IsBare {
+		n, err = client.GetNotificationData(n)
```

Comment:
`client.GetNotificationData` mutates the passed in `notification` and populates it with data, but does not set `IsBare` to false. Should it be setting it to false or is this unnecessary?

Reply:
It does not mutate notification, but it was not indeed re-setting `IsBare`. Fixed now (along with a bit of naming cleanup).

Reply:
I want to make sure I understand this. We're passing in a pointer to a struct, dereferencing the pointer, modifying the struct, and returning a pointer to the same struct. How is this not mutating the struct? It's not important for discussion of the PR, but I don't understand.